src/sentry/sentry_metrics/indexer/limiters/cardinality.py

```diff
@@ -0,0 +1,147 @@
+        ]
+    else:
+        raise ValueError(use_case_id)
+
+
+class InboundMessage(TypedDict):
+    # Note: This is only the subset of fields we access in this file.
+    org_id: int
+    name: str
+    type: str
+    tags: Mapping[str, str]
+
+
+class TimeseriesCardinalityLimiter:
+    def __init__(self, namespace: str, rate_limiter: CardinalityLimiter) -> None:
+        self.namespace = namespace
+        self.rate_limiter: CardinalityLimiter = rate_limiter
+
+    def check_cardinality_limits(
+        self, use_case_id: UseCaseKey, messages: Mapping[TMessageKey, InboundMessage]
+    ) -> CardinalityLimiterState[TMessageKey]:
+        request_hashes = defaultdict(set)
+        keys_to_remove = {}
+        for key, message in messages.items():
+            org_id = message["org_id"]
+            message_hash = hash_values(
+                [
+                    message["name"],
+                    message["type"],
+                    message["tags"],
+                ]
+            )
+            prefix = _build_quota_key(self.namespace, org_id)
+            keys_to_remove[prefix, message_hash] = key
+            request_hashes[prefix].add(message_hash)
+
+        requested_quotas = []
+        configured_quotas = _construct_quotas(use_case_id)
+        for prefix, hashes in request_hashes.items():
+            requested_quotas.append(
+                RequestedQuota(prefix=prefix, unit_hashes=hashes, quotas=configured_quotas)
+            )
+
+        timestamp, grants = self.rate_limiter.check_within_quotas(requested_quotas)
+
+        for grant in grants:
+            for hash in grant.granted_unit_hashes:
+                del keys_to_remove[prefix, hash]
+
+        return CardinalityLimiterState(
+            _cardinality_limiter=self.rate_limiter,
+            _use_case_id=use_case_id,
+            _requests=requested_quotas,
+            _grants=grants,
+            _timestamp=timestamp,
+            keys_to_remove=list(keys_to_remove.values()),
+        )
+
+    def apply_cardinality_limits(self, state: CardinalityLimiterState[Any]) -> None:
+        state._cardinality_limiter.use_quotas(state._requests, state._grants, state._timestamp)
+
+
+class TimeseriesCardinalityLimiterFactory:
+    """
+    The TimeseriesCardinalityLimiterFactory is in charge of initializing the
+    TimeseriesCardinalityLimiter based on a configuration's namespace and
+    options. Ideally this logic would live in the initialization of the
+    backends (postgres, cloudspanner etc) but since each backend supports
+    multiple use cases dynamically we just keep the mapping of rate limiters in
+    this factory.
+
+    [Copied from sentry.sentry_metrics.indexer.limiters.writes]
+    """
+
+    def __init__(self) -> None:
+        self.rate_limiters: MutableMapping[str, TimeseriesCardinalityLimiter] = {}
+
+    def get_ratelimiter(self, config: MetricsIngestConfiguration) -> TimeseriesCardinalityLimiter:
```

Comment:
@fpacifici 
It is now namespaced the same way as the writes limiter, but I wonder whether we want to actually namespace this rate limiter per-indexer-backend? Since we are protecting clickhouse more than the indexer here.

Reply:
That is right. We do not need that, at least not in the immediate.

Reply:
i tried to fix that by having a separate namespace config for the cardinality limiter.